minion/libwrapper.cpp

```diff
@@ -7,12 +7,188 @@
 
 #include "libwrapper.h"
 #include "minion.h"
+#include "command_search.h"
+#include "info_dumps.h"
+#include "system/minlib/exceptions.hpp"
+#include <iomanip>
 
 #ifdef LIBMINION
+
+// from minion.cpp
+void doStandardSearch(CSPInstance& instance, SearchMethod args);
+void finaliseModel(CSPInstance& instance);
+
+
 extern Globals* globals;
 
 void resetMinion() {
   delete globals;
   globals = new Globals();
 }
+
+ReturnCodes runMinion(SearchOptions& options, SearchMethod& args, ProbSpec::CSPInstance& instance, bool(*callback)(void)) {
+
+  ReturnCodes returnCode = ReturnCodes::OK;
+
+  /*
+   * Adapted from minion_main.
+   * Whereas minion_main takes in command line arguments, we take in Minion
+   * objects.
+   */
+
+  resetMinion();
+
+
+  // Redirect cout
+  // https://stackoverflow.com/questions/49462524/controlling-output-from-external-libraries
+  // https://stackoverflow.com/questions/4810516/c-redirecting-stdout
+
+  streambuf* oldCoutStreamBuf = cout.rdbuf();
+  ifstream logOutStream;
+  time_t rawtime;
+  time(&rawtime);
+
+  stringstream filenameStream;
+  filenameStream << "minion";
+  filenameStream << put_time(gmtime(&rawtime), "%Y-%m-%d-%H:%M:%S");
+  filenameStream << ".log";
+
+  logOutStream.open(filenameStream.str(),ios_base::app);
+  
+  cout.rdbuf(logOutStream.rdbuf());
+
+  // Pass error codes across FFI boundaries, not exceptions.
+  try {
+
+    // No parallel minion in library usage for now
+    // getParallelData();
+
+    getState().getOldTimer().startClock();
+
+    globals->callback = callback;
+    globals->options_m = new SearchOptions(options);
+
+    getOptions().printLine("# " + std::string(MinionVersion));
+    getOptions().printLine("# Git version: \"" + tostring(GIT_VER) + "\"");
+
+    GET_GLOBAL(global_random_gen).seed(args.randomSeed);
+     if(!getOptions().silent) {
+        cout << "#  Run at: UTC " << asctime(gmtime(&rawtime)) << endl;
+        cout << "# Input filename: " << getOptions().instance_name << endl;
+        getOptions().printLine("Using seed: " + tostring(args.randomSeed));
+        }
+      
+
+    Parallel::setupAlarm(getOptions().timeoutActive, getOptions().time_limit,
+                         getOptions().time_limit_is_CPUTime);
+
+    finaliseModel(instance);
+
+    // Output graphs, stats, or redump (will not return in these cases)
+    infoDumps(instance);
+
+    // Copy args into tableout
+    getTableOut().set("RandomSeed", tostring(args.randomSeed));
+    getTableOut().set("Preprocess", tostring(args.preprocess));
+
+    getTableOut().set("MinionVersion", -1);
+    getTableOut().set("TimeOut", 0); // will be set to 1 if a timeout occurs.
+    getState().getOldTimer().maybePrintTimestepStore(cout, "Parsing Time: ", "ParsingTime",
+                                                     getTableOut(), !getOptions().silent);
+
+    SetupCSPOrdering(instance, args);
+    BuildCSP(instance);
+
+    //TODO (nd60): how to replace this??
+    if(getOptions().commandlistIn != "") {
```

Comment:
just ignore CommandSearch. Long term, the minion interface should grow to provide the kind of functionality command search provides, but it will be done in an entirely different way.